workspaces/src/rpc/client.rs

```diff
@@ -2,41 +2,254 @@
 //       warnings about unstable API.
 #![allow(deprecated)]
 
+use std::collections::HashMap;
+
 use tokio_retry::strategy::{jitter, ExponentialBackoff};
 use tokio_retry::Retry;
 
+use near_crypto::{InMemorySigner, PublicKey, Signer};
+use near_jsonrpc_client::methods::query::RpcQueryRequest;
 use near_jsonrpc_client::{methods, JsonRpcClient, JsonRpcMethodCallResult};
-use near_primitives::transaction::SignedTransaction;
-use near_primitives::views::FinalExecutionOutcomeView;
+use near_jsonrpc_primitives::types::query::QueryResponseKind;
+use near_primitives::account::{AccessKey, AccessKeyPermission};
+use near_primitives::hash::CryptoHash;
+use near_primitives::transaction::{
+    Action, AddKeyAction, CreateAccountAction, DeleteAccountAction, DeployContractAction,
+    FunctionCallAction, SignedTransaction, TransferAction,
+};
+use near_primitives::types::{AccountId, Balance, Finality, FunctionArgs, Gas, StoreKey};
+use near_primitives::views::{AccessKeyView, FinalExecutionOutcomeView, QueryRequest};
 
-use crate::runtime::context::MISSING_RUNTIME_ERROR;
+use crate::rpc::tool;
 
-fn rt_current_addr() -> String {
-    crate::runtime::context::current()
-        .expect(MISSING_RUNTIME_ERROR)
-        .rpc_addr()
-}
+const DEFAULT_CALL_FN_GAS: Gas = 10000000000000;
+const ERR_INVALID_VARIANT: &str =
+    "Incorrect variant retrieved while querying: maybe a bug in RPC code?";
 
-fn json_client() -> JsonRpcClient {
-    JsonRpcClient::connect(&rt_current_addr())
+fn json_client(addr: &str) -> JsonRpcClient {
+    JsonRpcClient::connect(addr)
 }
 
-pub(crate) fn new() -> RetryClient {
-    RetryClient::new()
+/// A client that wraps around JsonRpcClient, and provides more capabilities such
+/// as retry w/ exponential backoff and utility functions for sending transactions.
+pub struct Client {
+    rpc_addr: String,
 }
 
-pub struct RetryClient;
-
-impl RetryClient {
-    fn new() -> Self {
-        Self {}
+impl Client {
+    pub(crate) fn new(rpc_addr: String) -> Self {
+        Self { rpc_addr }
     }
 
-    pub(crate) async fn call<M: methods::RpcMethod>(
+    pub(crate) async fn query<M: methods::RpcMethod>(
         &self,
         method: &M,
     ) -> JsonRpcMethodCallResult<M::Result, M::Error> {
-        retry(|| async { json_client().call(method).await }).await
+        retry(|| async { json_client(&self.rpc_addr).call(method).await }).await
+    }
+
+    async fn send_tx_and_retry(
+        &self,
+        signer: &InMemorySigner,
+        receiver_id: AccountId,
+        action: Action,
+    ) -> anyhow::Result<FinalExecutionOutcomeView> {
+        send_batch_tx_and_retry(self, signer, receiver_id, vec![action]).await
+    }
+
+    pub async fn call(
```

Comment:
I think this is not visible outside of the crate. If that's the case, I suggest making it as `pub(crate)`. What we are developing is a public Rust API, so it is super-duper important to make sure that actually `pub` things have correct signatures, and for that, it's quite useful to tell at a glance whether the stuff is actually visible from outside of the crate or not.

Reply:
I was thinking of giving this out to everyone to be able to make their own custom Networks later on (similarly w/ all the traits), but for now I'll make it all crate private just so we don't confuse anyone and see what people want out of it. And then maybe later expose it